### PR TITLE
Thurloe's got Swagger

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

**Comment by mcovarr** [comment]
@coreone do we need DevOps magic beyond the `.travis.yml` I just added to get Travis set up for Thurloe?

**Comment by coreone** [comment]
Yes...I needed to go click a button.  Should be Travis-ized now:

https://travis-ci.org/broadinstitute/thurloe

**Comment by mcovarr** [comment]
@coreone This can definitely wait until next week, but it looks like we might need a Coveralls repo set up too.

**Comment by mcovarr** [comment]
@kshakir second reviewer ( @cjllanwarne first reviewer )

File: src/test/scala/thurloe/service/ThurloeServiceSpec.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ class ThurloeServiceSpec extends FunSpec with ScalatestRouteTest {
 
   describe("The Thurloe Service") {
     it("should allow key/value pairs to be set") {
-      Post(uriPrefix, ukvp1) ~> thurloeService.routes ~> check {
+      Post(uriPrefix, ukvp1) ~> thurloeService.keyValuePairRoutes ~> check {
```
Comment (cjllanwarne) [inline src/test/scala/thurloe/service/ThurloeServiceSpec.scala:30]: I had these being scanned by the entire routing sequence to check the routing was working holistically.

File: src/main/resources/application.conf [hunk 2]
```diff
@@ -1,4 +1,18 @@
...
+  api {
+    info = "Thurloe"
+    version = "0.02" // TODO: Read at runtime?
+    contact = "dsde@broadinstitute.org"
+    description = "all your key/value pairs are belong to us"
+    license = "BSD"
+    licenseUrl = "http://opensource.org/licenses/BSD-3-Clause"
+    termsOfServiceUrl = "http://www.github.com/broadinstitute/thurloe"
+  }
+  ui.version = "2.1.1"
```
Comment (cjllanwarne) [inline src/main/resources/application.conf]: Is this copy/paste from another project? What does 2.1.1 represent here?

File: src/main/resources/swagger/thurloe.yaml [hunk 3]
```diff
@@ -1,7 +1,8 @@
 swagger: "2.0"
 info:
-  title: Thurloe Service
-  version: 0.1
+  title: Thurloe
+  description: All your key/value pairs are belong to us
```
Comment (cjllanwarne) [inline src/main/resources/swagger/thurloe.yaml:4]: :+1: (Just for this line)

File: src/main/scala/thurloe/service/ThurloeService.scala [hunk 4]
```diff
@@ -38,14 +40,14 @@ trait ThurloeService extends HttpService {
         case Failure(e) =>
           respondWithStatus(StatusCodes.InternalServerError) {
             complete {
-              s"Oops! $e"
+              s"$Interjection! $e"
```
Comment (cjllanwarne) [inline src/main/scala/thurloe/service/ThurloeService.scala]: Should the ! be part of the interjection? Maybe I'm overthinking this.

File: src/main/scala/thurloe/service/ThurloeService.scala [hunk 5]
```diff
@@ -115,12 +117,16 @@ trait ThurloeService extends HttpService {
...
             }
           }
       }
     }
   }
 
-  val routes = getRoute ~ getAllRoute ~ setRoute ~ deleteRoute
+  def yamlRoute = path("swagger" / Yaml) {
```
Comment (cjllanwarne) [inline src/main/scala/thurloe/service/ThurloeService.scala]: Have "swagger" as a val (alongside "ThurloePrefix") for easy-to-find-and-change

**Comment by cjllanwarne** [comment]
:+1:

File: build.sbt [hunk 6]
```diff
@@ -14,6 +14,7 @@ val akkaV = "2.3.12"
 
 libraryDependencies ++= Seq(
   "org.scala-lang" % "scala-reflect" % "2.11.7",
+  "com.gettyimages" %% "spray-swagger" % "0.5.1",
```
Comment (kshakir) [inline build.sbt]: We no longer use `spray-swagger`. This is leftover code cruft. Will point out the other parts below.

File: src/main/resources/application.conf [hunk 7]
```diff
@@ -1,4 +1,18 @@
...
+  baseUrl = ""
+  api {
+    info = "Thurloe"
+    version = "0.02" // TODO: Read at runtime?
+    contact = "dsde@broadinstitute.org"
+    description = "all your key/value pairs are belong to us"
+    license = "BSD"
+    licenseUrl = "http://opensource.org/licenses/BSD-3-Clause"
+    termsOfServiceUrl = "http://www.github.com/broadinstitute/thurloe"
+  }
```
Comment (kshakir) [inline src/main/resources/application.conf]: This entire api block is not required.

File: src/main/scala/thurloe/webservice/SwaggerConfigHttpService.scala [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package thurloe.webservice
+
+import com.gettyimages.spray.swagger.SwaggerHttpService
```
Comment (kshakir) [inline src/main/scala/thurloe/webservice/SwaggerConfigHttpService.scala]: This `SwaggerHttpService` again isn't used. We're not even using its [`final def routes`](https://github.com/gettyimages/spray-swagger/blob/v0.5.1/src/main/scala/com/gettyimages/spray/swagger/SwaggerHttpService.scala#L60) anywhere in the trait, and they were removed from consumer's composed route [here](https://github.com/broadinstitute/thurloe/pull/14/files#diff-9da5137ca5614efd1f2591aac06eb049L26).

File: src/main/scala/thurloe/webservice/SwaggerConfigHttpService.scala [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  // NOTE: These two values in the conf must be kept in sync with changes to build.sbt
+  // Validated by SwaggerConfigHttpServiceSpec
+  override lazy val apiVersion: String = swaggerConfig.getString("api.version")
+  override lazy val swaggerUiVersion: String = swaggerConfig.getString("ui.version")
+
+  override lazy val docsPath: String = optString("docsPath", super.docsPath)
+  override lazy val swaggerVersion: String = optString("swaggerVersion", super.swaggerVersion)
+  override lazy val uiPath: String = optString("ui.path", super.uiPath)
+
+  override lazy val apiInfo: Option[ApiInfo] =
```
Comment (kshakir) [inline src/main/scala/thurloe/webservice/SwaggerConfigHttpService.scala]: Not needed anymore.

Comment (kshakir) [inline src/main/scala/thurloe/webservice/SwaggerConfigHttpService.scala]: Several of the above properties will be marked as not required also once `spray-swagger` is removed. All of `spray-swagger` has been superseded by the use of static yaml instead.

**Comment by kshakir** [comment]
FYI, feel free to clean this up, but I've also refactored a version of the swagger service into lenthall. It's currently awaiting completion of review here: broadinstitute/lenthall#3

Post review it'll be automatically published and available for use in thurloe.

**Comment by mcovarr** [comment]
@kshakir Thanks for looking this over.  Given broadinstitute/lenthall#3, I'm thinking it's probably a better to wait for that to become available rather than spend time cleaning this up.

File: src/main/resources/application.conf (hunk 2 above)
Comment (mcovarr) [inline src/main/resources/application.conf]: Yes this is all horrendous copy/paste.

File: src/test/scala/thurloe/service/ThurloeServiceSpec.scala (hunk 1 above)
Comment (mcovarr) [inline src/test/scala/thurloe/service/ThurloeServiceSpec.scala:30]: I'll see if I can restore this after Lenthall refactoring.

**Comment by kshakir** [comment]
Example usage of lenthall's `spray-can` and `swagger-ui` utils (and ripping out of `spray-swagger`) are available in [this commit](https://github.com/broadinstitute/cromwell/commit/1bbdfc5b620de88cbe9f9b3b42b7a4dc6be25503) of this PR: broadinstitute/cromwell#237

File: src/main/resources/application.conf (hunk 2 above)
Comment (kshakir) [inline src/main/resources/application.conf]: Post lenthall-izing, the `swagger` block should look like:

``` yaml
swagger {
  docsPath = "swagger/thurloe.yaml"
  uiVersion = "2.1.1"
}
```

File: src/test/scala/thurloe/service/ThurloeServiceSpec.scala (hunk 1 above)
Comment (mcovarr) [inline src/test/scala/thurloe/service/ThurloeServiceSpec.scala:30]: Actually I'm not sure I changed anything here other than the name of the member from `routes` to `keyValuePairRoutes`; am I missing a more substantive change here?

**Comment by mcovarr** [comment]
Now! with Lenthall :tm: technology!

File: .coveralls.yml [hunk 10]
```diff
@@ -0,0 +1 @@
+service_name: travis-ci
```
Comment (kshakir) [inline .coveralls.yml]: Don't think this file does anything, but doesn't hurt either?

**Comment by kshakir** [comment]
One question/comment, but go for merge either way! :+1:

**Pull request merged**